Ensure plug-in C modules don't mix with target ones.

dschaefer2 opened the pull request:
When a Swift module depends on a C library that is also a dependent of plug-in that the module depends on, we were seeing includes of the module map for both versions of that library, host and target, causing types to be redefined.

This change filters out the plug-in version of these dependencies.

dschaefer2 commented:
Still have tests to write.

MaxDesiatov commented on Sources/Build/BuildPlan/BuildPlan+Swift.swift:
```diff
@@ -26,6 +26,10 @@ extension BuildPlan {
                 guard case let .clang(target)? = description else {
                     throw InternalError("unexpected clang target \(underlyingTarget)")
                 }
+                // Only if the targets are for the same destination, i.e. not a plugin for a target target
+                guard target.buildParameters.destination == swiftTarget.buildParameters.destination else {
+                    break
+                }
```
formatting nit: IMO more readable with visual separation, and makes it easier to navigate with `{`/`}` when using Vi keybindings 😉 

```suggestion
                }

                // Only if the targets are for the same destination, i.e. not a plugin for a target target
                guard target.buildParameters.destination == swiftTarget.buildParameters.destination else {
                    break
                }

```

MaxDesiatov reviewed (commented):
> we were seeing includes of the module map for both versions of that library, host and target, causing types to be redefined.

Would you mind adding a test case that reproduce that situation?

dschaefer2 commented:
@swift-ci please test

dschaefer2 commented:
Updated to push this all the way to the dependency walking so we never get plug-in rooted dependencies in this part of the planning. And added a test.

Also missed adding the issue link #8060 which points at the @DougGregor's repo where I first reproduced this.

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -6736,4 +6736,68 @@ final class BuildPlanTests: XCTestCase {
         }
     }
+
+    func testHostTargetDontCross() async throws {
```
naming nit: `Cross` could refer to cross compilation here, but I thought we're actually referring to overlapping builds?

```suggestion
    func testHostTargetDontOverlap() async throws {
```

dschaefer2 replied:
It's referring to crossing the streams. Been watching too much Ghostbusters lately :p.

dschaefer2 replied:
Renamed to better reflect.

dschaefer2 commented:
@swift-ci please test

dschaefer2 commented:
@swift-ci please test windows

xedin commented on Sources/Build/BuildPlan/BuildPlan+Swift.swift:
```diff
@@ -20,12 +20,13 @@ extension BuildPlan {
         // We need to iterate recursive dependencies because Swift compiler needs to see all the targets a target
         // depends on.
-        for case .module(let dependency, let description) in swiftTarget.recursiveDependencies(using: self) {
+        for case .module(let dependency, let description) in swiftTarget.recursiveLinkDependencies(using: self) {
```
I think instead of a special method that is filtering out plugins what you want to do is check whether the dependency targets the same destination and if not skip it, that would filter out plugins/macros and other things for "target" destination.

dschaefer2 replied:
I did that originally but felt that was hiding the real issue of including plug-in dependencies in my list of recursive dependencies when I don't want them when constructing the command line arguments for my module build.

xedin replied:
Maybe add a optional destination as a parameter to recursive dependencies method then, it would be easier to understand than having to navigate multiple slightly different methods.

xedin replied:
Or maybe it would be okay to adjust `recursiveDependencies` to always respect destination of the requesting module I don't think we ever want to have things from different destinations used anyway, it's not only plugins though it's also applicable to macros.

dschaefer2 replied:
I've been hesitant to rely on destinations. Could a plug-in tool depend on another plug-in tool? Wouldn't both be host?

xedin replied:
Yes, both would be host, just like macros. There are other places in Build which filter based on the destination so it won’t be unique.

xedin replied:
Btw, note that plugins actually have a special description not `SwiftModuleBuildDescription`.

dschaefer2 commented:
@swift-ci please test

dschaefer2 commented:
@swift-ci please test windows

dschaefer2 commented:
Fixed this for real https://github.com/swiftlang/swift-package-manager/pull/8524
